### DISCUSSION TITLE
Tags/semantic decorators

### DISCUSSION THREAD

**Discussion by IS4Code** [dotnet/csharplang#3768] [Language Ideas]
This proposal is inspired in part by F# and its units of measurement feature, and in part by [Parse, don’t validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).

A "tag" is a kind of a type that can be attached to any type or value on a per-use basis, and serves to further distinguish values of any type. The main purpose of tags is to introduce lightweight "decorators" into the type system that don't affect the runtime behaviour of the code, but allow the static type checking to guarantee certain invariants are upheld, similarly to non-nullable reference types.

Syntactically, a tag is a combination of a static class, an interface, and a shape. However, instead of restricting functionality of types it is used in combination with, like an interface, it can extend any operation on a normal type by means of specifying the result tag.

As a simple example:
```cs
public tag Verified; // any piece of data tagged with this must have been processed by some user-defined verification code

public static void SendData(Verified(string) data) { ... }

public static Verified(string) Verify(string str)
{
    if(str.Length > 0) return Verified(str);
    throw new ArgumentException(nameof(str));
}

SendData("abc"); // does not work (error/warning)
SendData(Verify("abc")); // fine
```

A tagged type is considered its own type in C#, acting as a derived/more specific type when compared with its untagged version. Specifically, a tagged type can be used as a return type, parameter/variable type, as a generic argument or constraint, or even the base type. In contrast, a tag itself, without being applied to anything, is only valid as the operand of `typeof`, or when inherited by another tag. Method overloads distinguished solely by the tags of their parameters should be allowed.

A tag can be derived from another tag, a specific non-static type (class, struct, delegate, enum, interface), a tagged type, or a shape, all permitted simultaneously. When derived from a tag, it specifies implication; when derived from anything else, it indicates a restriction (the to-be-tagged type must be assignable/conform to the restricted type).
```cs
public tag MoreVerified : Verified, string;

_ = MoreVerified(1); // MoreVerified(int) is invalid; int is not string
SendData(MoreVerified("abc")); // MoreVerified(string) is Verified(string)
```

A type tagged with the same tag twice is different from when it is tagged only once, also the order of applied tags matters.

#### Operations

A tag can also specify its members. but any member that is an instance member or an operator method is considered special. Non-special (thus static) members work like on any other type; they are accessed and invoked via the tag itself. Special members are not permitted to have a body, and any type used therein can be a tag.

When invoking members on a tagged type, the lookup works the same as for an untagged type. However, when the tag itself has a member that matches the invoked member, it can add tag information to the type of the member.

```cs
public tag NonZero
{
    public static NonZero operator*(NonZero a, NonZero b);
    public static NonZero operator-(NonZero a);
}

public tag Positive : NonZero
{
    public static Positive operator+(Positive a, Positive b);
    public static Positive operator*(Positive a, Positive b);
    public static Negative operator*(Positive a, Negative b);
    public static Negative operator*(Negative a, Positive b);
    public static NonZero operator*(Positive a, NonZero b);
    public static NonZero operator*(NonZero a, Positive b);
    public static Negative operator-(Positive a);
}

public tag Negative : NonZero
{
    public static Negative operator-(Negative a, Negative b);
    public static Positive operator*(Negative a, Negative b);
    public static Negative operator*(Negative a, Positive b);
    public static Negative operator*(Positive a, Negative b);
    public static NonZero operator*(Negative a, NonZero b);
    public static NonZero operator*(NonZero a, Negative b);
    public static Positive operator-(Negative a);
}

public tag NonEmpty
{
    public Positive Length { get; }
    public Positive Count { get; }
}
```

`Positive(3) + Positive(5)` is `Positive(int)`, but `Positive(3) - Positive(5)` is simply `int`. As long as the tag is applied in a semantically safe way, all its subsequent uses are also safe.

A private pseudo-constructor syntax could be used to restrict the accessibility of the tag when applied to values:
```cs
public tag NonEmpty
{
    private NonEmpty(); // can only be applied to values from within itself

    public Positive Length { get; }
    public Positive Count { get; }
    
    public static NonEmpty(T) Ensure(T arg) where T : Array
    {
        if(arg.Length == 0) throw new ArgumentException(nameof(arg));
        return NonEmpty(arg);
    }
}
```

#### Generics
It makes sense to allow tags to be generic. It also makes sense to allow not only tagged types as generic arguments, but also tags themselves, when the generic parameter specifies it.

```cs
public class Container<tag T> where T : int, new()
{
    public T(int) Field = T(0);
}
```

In the case of a tag parameter, the constraints are always necessary. `int` specifies the tag can be applied to `int` (not restrictively, but only permissively), and `new()` specifies its "constructor" is visible and thus it can be freely applied to values.

### Implementation
In signatures, it is viable to emit tags as either `modreq` or `modopt`. At runtime, it seems better not to represent tags at all. A considerable option would be to represent `NonZero(int)` as `NonZero<int>` and `NonZero(1)` as `new NonZero<int>(1)` for a new type similar to `ValueTuple<int>`, but this violates the inheritance chain and makes `(object)NonZero(1) is int` fail.

### Alternatives
There are several other options for the syntax, from which I have chosen the one that seemed both easy to grow accustomed to, and easy to parse.
```
NonZero(int)
NonZero(1)
// unnatural
NonZero[int]
NonZero[1]
// generic tags won't be a thing with this
NonZero<int>
NonZero<1>
// possibly ambiguous
NonZero int
NonZero 1
// could be mistaken for a label
NonZero: int
NonZero: 1
```

**Comment by Flutterish** [comment]
Are you sure youre not just describing [Shapes](https://github.com/dotnet/csharplang/issues/164)?

**Comment by IS4Code** [comment]
@Flutterish I thought I made a pretty good distinction there.

Shapes specify how classes should behave, i.e. what members they should have. Tags *extend* how *instances* of classes behave when invoked or combined with others by adding tags to the return type.

Shapes encode compile-time invariants about the type. Tags encode runtime invariants about the *value*.

The shape of a type is specified just once as a constraint. The tag of a type is flexible and can be added or removed wherever the type is used, based on the value, which we then say to be *parsed* as the tagged type when the runtime invariants the tag specifies are checked.

Shapes are compiled as interfaces and this way they allow the runtime to call methods on the witness. Tags remain static classes and they are not handled by the runtime at all, only present in metadata/signatures (similarly to static classes in [System.Runtime.CompilerServices](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices?view=netframework-4.8)).

"Applying" a shape to a type requires compatibility of the members. Applying a tag to a type only requires it matching the explicit constraints specified by the tag itself (possibly a shape). If a tag has any members, they don't have to be present on the to-be-tagged type, it only specifies what invariants can be implied for members or results of operations.

In general, a shape is a specific condition that an input type must meet, and this allows you to call its methods. A tag is an implied condition a value must meet, and prevents you from using "invalid" values without spilling checks all over the place.

You are probably using a tag without even realizing it, and the tag is "NonNullable". The distinction between `object` and `object?` is not a shape, rather an additional information about the values of the former. In the language of this proposal, it would be `NonNullable(object)` and `object`. This proposal allows encoding *any* user-defined information about a value of a particular type - a reference not being null, an object not being disposed, a struct not being initialized to its default state, a collection not being empty, a string matching a particular format etc.

The only slight issue is that tags are "safe" only when encoding information about a value that never changes once checked, but still, if you are using immutable data structures, you are safe and free from checks. And `NonEmpty(string)` or perhaps `NonEmpty(Span<int>)` is still useful.

In a sense, tags are generalized explicit contracts when treated as a part of the type system.

**Comment by 333fred** [comment]
This sounds a lot like roles (https://github.com/dotnet/csharplang/issues/1711) to me.

**Comment by IS4Code** [comment]
@333fred Good point! It seems the declaration of a tag is essentially a generic role, but I am unsure about the members:
```cs
public role NonEmpty<T> of T
{
    public override Positive<> Length => base.Length;
    public override Positive<> Count => base.Count;
}
```
The point was either one of `Count` or `Length` would apply in cases where the type had inconsistent member names, but now `this` is of opaque type `T` and so `this.Length` would have to be generated by the compiler for every `T` used with `NonEmpty<>`, essentially turning it to a template. Tags don't need that since their special members are only there for passive extension of already existing members.

**Comment by Reinms** [comment]
I think the comparison with nullable is a lot closer than with roles. They are less a type system and more a way to augment static type checking, which lands them pretty close to source generator territory.

I have mixed feelings about that, on one hand this is something may be possible to create right now. On the other hand, it seems like potentially a lot of headache, and source generators sit pretty high on my list of "Things I should learn and want to learn, but get put off by the entry barrier"

Perhaps it would be worthwhile to add "tags-like system" to the list of common use cases for source generators and investigate potential roadblocks to doing so and if there are any ways to streamline their creation?

For what it's worth, tags would be incredibly useful to me currently. I'm in the midst of writing a compile time type and stack checked abstraction layer over reflection.emit that is implemented entirely through generics, and it is quite the headache at times.

**Comment by jmarolf** [comment]
This is a weird pull, but this use case most reminds me of Ada's [subtypes](https://en.wikibooks.org/wiki/Ada_Programming/Type_System#Creating_subtypes) feature. I could see being able to restrict a type (but not necessarily imply an inheritance relationships) being useful. I do agree with @333fred, I think the roles proposal covers this use case and more. I would imagine that the design of "roles" could be changed so, at minimum, it would behave like "type tags". I think that is actually the main use case for roles when combined with pattern matching and (hopefully) discriminated unions.

Related discussions:
- [Roles](https://github.com/dotnet/csharplang/issues/1711) 
- [Shapes](https://github.com/dotnet/csharplang/issues/164)
	- [Type Classes](https://github.com/dotnet/csharplang/issues/110)
	- [Extension Everything](https://github.com/dotnet/roslyn/issues/11159)
- [Discriminated Unions](https://github.com/dotnet/csharplang/issues/113)
	- [Discriminated Unions Discussion](https://github.com/dotnet/csharplang/issues/399)
	- [Discriminated Unions Discussion Conintued..](https://github.com/dotnet/csharplang/issues/3760)
